app/Dockerfile

```diff
@@ -0,0 +1,63 @@
+# from https://bun.sh/guides/ecosystem/docker, with modifications
+# to run a hot-reloading development server
+
+# use the official Bun image
+# see all versions at https://hub.docker.com/r/oven/bun/tags
+FROM oven/bun:1 AS base
+WORKDIR /app
+
+
+# -----------------------------------------------------------
+# install dependencies for dev and prod into temp directories
+# this will cache them and speed up future builds
+FROM base AS install
+
+COPY package.json bun.lockb /temp/dev/
+RUN --mount=type=cache,target=/tmp/bun \
+    cd /temp/dev/ && \
+    bun install --cache-dir /tmp/bun --frozen-lockfile
+
+# install with --production (exclude devDependencies)
+COPY package.json bun.lockb /temp/prod/
+RUN --mount=type=cache,target=/tmp/bun \
+    cd /temp/prod && \
+    bun install --cache-dir /tmp/bun --frozen-lockfile --production
+
+
+# -----------------------------------------------------------
+# copy node_modules from temp directory
+# then copy all (non-ignored) project files into the image
+FROM base AS prerelease
+COPY --from=install /temp/dev/node_modules node_modules
+COPY . .
+
+# [optional] tests & build
+ENV NODE_ENV=production
+RUN bun test
+RUN bun run build
+
+
+# -----------------------------------------------------------
+# copy node_modules from dev stage, copy entire app
+# source into the image
+FROM base AS dev
+COPY --from=install /temp/dev/node_modules node_modules
+COPY . .
+# run the app in hot-reloading development mode
+# (this requires a new script, container-dev, in package.json that sets up vite
+# to accept connections on any interface, e.g. from outside the container, and
+# to always run on port 5713)
+CMD ["bun", "run", "container-dev"]
+
+
+# -----------------------------------------------------------
+# copy production dependencies and source code into final image
+FROM base AS release
+COPY --from=install /temp/prod/node_modules node_modules
+COPY --from=prerelease /app/main.tsx .
+COPY --from=prerelease /app/package.json .
+
+# run the app in production mode
+USER bun
+EXPOSE 3000/tcp
+ENTRYPOINT [ "bun", "run", "main.ts" ]
```

Comment:
I forgot to comment on and test this. While this may work, my feeling is that we should just the built-in vite `build` and `preview` scripts, if the intention is to run the production build of the frontend. I think that will be more reliable.

Reply:
Actually, what does this intend to do. It's `main.tsx` in the copy command, and running `main.ts` (not sure where this comes from)? And it's running with bun, when this code is intended to be run in a browser. 

I guess you're already building the frontend in --prerelease and taking artifacts from there, but shouldn't it be taking the `/dist` artifacts, and hosting them on a server?

Reply:
Isn't `bun run` just running entries from `scripts` in `package.json`, which already reference `vite`? I don't see how it's any more reliable to circumvent that and invoke `vite` directly. I'm fine with changing it, but I want to understand.

Reply:
> ENTRYPOINT [ "bun", "run", "main.ts" ]

Yes, here you're running `bun run main.ts`, right? Which is  saying run the `main.ts` script in `package.json`, which there is none as far as I see. And `main.ts` isn't a file anywhere as far as I can see either.

I'm not super clear on the Docker commands here, but I think we want something like this:

```dockerfile
COPY --from=install /temp/prod/node_modules node_modules
COPY --from=prerelease /app/dist

# run the app in production mode
USER bun
EXPOSE 4173/tcp
ENTRYPOINT [ "bun", "run", "preview" ]
```

Need to make use of the resulting `/dist` folder from `bun run build` in the "prerelease" section, and run `bun run preview` to vite serve that folder on a local server.

Reply:
Sorry, my comment got mixed in with your later one. To clarify, the code from the [original bun Dockerfile](https://bun.sh/guides/ecosystem/docker) is supposed to run a "production" instance of a webserver that's hosting the frontend app. Since we're serving the app as a bundle from a reverse http proxy in production, we don't want that; I changed it to instead just build the app as a bundle when you use the '"release" target.